ch4/ofi: Refactor RMA window setup and enable scalable EP for RMA window

shawnccx opened the pull request:
This PR contains two major changes:
- Refactor RMA window setup: 
OFI_win_init function is split it into a few parts including per window sync, init shared context, init global context and set fi info for RMA capabilities.
- Enable scalable endpoint for RMA window: 
Scalable endpoint is enabled for RMA window to support multiple transmit contexts. When all transmit contexts are used up on scalable EP, it will fall back to use shared transmit context or global endpoint.

hajimefu commented:
@minsii As we agreed at #2934, this PR includes the modified version of "Refactor RMA window setup".

shawnccx commented:
@raffenet Can you help us kick off jenkins tests? thanks.

minsii commented:
test:jenkins/ch4/all

shawnccx commented:
Just rebased this PR to solve conflicts in master branch. Any comments? 
Or probably we can discuss this in next week's hackathon.

minsii commented:
@shawnccx : Sorry that I did not check this PR yet. I am tied up with some deadlines this week. Let us discuss this at hackathon.

shawnccx commented:
Modified commits, last commit is changed to: remove `per-window sync` CVAR and enable it by default when using scalable endpoint and shared transmit context.

minsii commented on src/mpid/ch4/netmod/ofi/ofi_win.h:
```diff
@@ -149,6 +149,76 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_win_set_per_win_sync(MPIR_Win * win)
+        /* Specify number of transmit context according user input and provider limit. */
+        MPIDI_Global.max_rma_tx_cnt = MPL_MIN(MPIDI_Global.prov_use->domain_attr->max_ep_tx_ctx,
+                                              MPIR_CVAR_CH4_OFI_MAX_RMA_SEP_CTX);
+        finfo->ep_attr->tx_ctx_cnt = MPIDI_Global.max_rma_tx_cnt;
```
It looks like the number of available transmit contexts in the scalable EP is only `1` by default. Can we set the default value of `MPIR_CVAR_CH4_OFI_MAX_RMA_SEP_CTX ` to something like "auto", thus `prov_use->domain_attr->max_ep_tx_ctx` number of transmit contexts can be used ?

shawnccx replied:
By default, current RMA path only uses `1` transmit context per window. So I just don't want to exhaust system resource by using `prov_use->domain_attr->max_ep_tx_ctx`.

minsii replied:
The scalable EP is shared by all windows, isn't it ? 
For example, only the first window can get a transmit context, then the second window will fail at `utarray_back(MPIDI_Global.rma_sep_idx_array)` . Please let me know if my understanding is wrong.

shawnccx replied:
Correct. Currently, `rma` and `pt2pt` are in the same domain. Take an example of `fi_info` in `psm2`:
```
        tx_ctx_cnt: 125
        rx_ctx_cnt: 125
        max_ep_tx_ctx: 128
        max_ep_rx_ctx: 128
```
And in `sockets`:
```
        tx_ctx_cnt: 16
        rx_ctx_cnt: 16
        max_ep_tx_ctx: 16
        max_ep_rx_ctx: 16
```
Making it as `1` is not good idea for window2, but set as maximum value exhausts all tx in the same domain.

minsii replied:
If set `finfo->ep_attr->tx_ctx_cnt= max_ep_tx_ctx ` , are all `max_ep_tx_ctx ` number of contexts reserved at the SEP creation time ? 

In current design, by default, the first window will create the global SEP with 1 `tx_ctx`, and then the second window will create the global shared context and share it with all later windows. This is misleading and the SEP with single `tx_ctx` is not useful here.

Since the SEP option is not useful unless the user specifies `MPIR_CVAR_CH4_OFI_MAX_RMA_SEP_CTX >1`, can we just disable SEP by default ?

shawnccx replied:
For first question, I think yes (as far as I can tell is from libfabric man page).

My original thought is to test `tx` in `sep` can work for RMA. It is hard to tell how many tx per sep we should set at this moment.  If you have multiple windows, then num of tx should be larger than number of window to be consistent.

I can turn sep off by default.

@hajimefu Any comments?

hajimefu replied:
@shawnccx I'm okay with turning it off by default, as long as it can be enabled via a CVAR.

minsii commented on src/mpid/ch4/netmod/ofi/ofi_types.h:
```diff
@@ -362,6 +363,7 @@ typedef struct {
     size_t rma_iov_limit;
     int max_ch4_vnis;
+    int max_rma_tx_cnt;         /* Max number of transmit context on one RMA scalable EP */
     size_t max_order_raw;
```
`max_rma_tx_cnt ` is used only for scalable EP, can we change the parameter name to something like `max_rma_sep_tx_cnt` so it is clear ?

minsii commented on src/mpid/ch4/netmod/ofi/ofi_types.h:
```diff
@@ -345,6 +345,7 @@ typedef struct {
     struct fid_cq *p2p_cq;
     struct fid_cntr *rma_cmpl_cntr;
     struct fid_stx *stx_ctx;    /* shared TX context for RMA */
+    struct fid_ep *rma_sep;     /* dedicated scalable EP for RMA */
```
`stx_ctx` is used only in RMA, maybe we want to change its name to `rma_ stx_ctx`.

shawnccx commented:
By default, socket provider turns `SEP` and `STX` on, they are defined in `src/mpid/ch4/netmod/ofi/ofi_capability_sets.h`.

shawnccx commented on src/mpid/ch4/netmod/ofi/ofi_win.h:
```diff
@@ -149,6 +149,76 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_win_set_per_win_sync(MPIR_Win * win)
+    MPIDI_OFI_CALL(fi_close(&MPIDI_OFI_WIN(win).ep->fid), epclose);
+    utarray_push_back(MPIDI_Global.rma_sep_idx_array, &(MPIDI_OFI_WIN(win).sep_tx_idx),
+                          MPL_MEM_RMA);
+    mpi_errno = MPI_ERR_OTHER;
```
This is available index of tx, we don't need to reset to `-1`. `-1` is an invalid value for tx index.

shawnccx replied:
you r correct, ignore my previous comment. I thought in reuse mode.

minsii commented on src/mpid/ch4/netmod/ofi/ofi_win.h:
```diff
@@ -111,6 +111,209 @@ static inline int MPIDI_OFI_win_allgather(MPIR_Win * win, void *base, int disp_u
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_OFI_WIN_INIT_SEP);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
```
The failure handling in `MPIDI_OFI_win_init_sep` is unclear to me. In current code:
1. If `fi_scalable_ep()` fails, `MPIDI_Global.rma_sep` will be NULL and `rma_sep_idx_array` will be NULL.
2. If `fi_scalable_ep_bind()` fails, `rma_sep` will not be NULL but `rma_sep_idx_array` will be NULL. However, `rma_sep_idx_array` will be illegally accessed in a later window creation call.
3. If `utarray_new()` fails (unlikely), no failure handling at all.
4. If `index == NULL`, which means no available `tx_ctx`, it returns `MPI_ERR_OTHER`.
5. If `fi_tx_context()` fails, just return and no `index` is popped.
6. If `fi_ep_bind()` fails, just return but `index` is popped.
7. If `MPIDI_OFI_win_set_per_win_sync` or `fi_enable` fails, the opened tx will be closed and `index` will be pushed back.

The concerns are:
- In 1, the SEP creation will be re-tried in every later window. This might be OK since the SEP resource might become available at a later time ?
- 2 and 3 need to be fixed
- In 4, "no available tx" is a valid case, we should not set an MPI error code here (it is not handled by caller routine though). We might want to use a flag for a "valid failure" ?
- 6 needs to be fixed: we should push `index` back.

It will be better if the code can clearly define the "valid failure" (e.g., no available resource) and "system failure" cases. The **MPI_ERR** code should be used only at "system failure", which might cause undefined behavior at continued execution; all "valid failure" cases should be handled by runtime properly.

shawnccx replied:
if `utarray_new()` fails, it just exists with `-1`. no failure handling can be added here.

shawnccx commented:
Revised patch, including:
- Fix var naming;
- Disable sep for RMA by default (setting sep_ctx CVAR to 0 by default)
- Fix error handling logic in `win_sep_init`: tx index in win is reset to -1; available index is pushed back into pool; two error flags are introduce in OFI level to replace MPI_ERR_OTHER; add error handle case for every possible OFI call in `win_sep_init`.

minsii commented on src/mpid/ch4/netmod/ofi/ofi_win.h:
```diff
@@ -111,6 +111,249 @@ static inline int MPIDI_OFI_win_allgather(MPIR_Win * win, void *base, int disp_u
+    }
+
+    MPIDI_OFI_CALL(fi_ep_bind(MPIDI_OFI_WIN(win).ep, &MPIDI_Global.rma_stx_ctx->fid, 0), bind);
+    MPIDI_OFI_CALL(fi_ep_bind(MPIDI_OFI_WIN(win).ep,
```
If `fi_ep_bind ` or `fi_enable` fails, the `MPIDI_OFI_WIN(win).ep` is overwritten in `MPIDI_OFI_win_init_global`, and will not be closed at `MPIDI_NM_mpi_win_free`.

shawnccx replied:
Okay, I will add error handling in `sxt_init` similar to `sep_init`.

minsii commented on src/mpid/ch4/netmod/ofi/ofi_win.h:
```diff
@@ -111,6 +111,249 @@ static inline int MPIDI_OFI_win_allgather(MPIR_Win * win, void *base, int disp_u
+        if (ret < 0) {
+            MPL_DBG_MSG(MPIDI_CH4_DBG_GENERAL, VERBOSE, "Failed to activate endpoint.\n");
+            mpi_errno = MPIDI_OFI_EPERROR;
+            goto fn_fail;
```
1. It is a little risky that you define `MPIDI_OFI_EPERROR` and `MPIDI_OFI_ENAVAIL` in OFI and assume they are not equal to `MPI_SUCCESS`. Perhaps it would be better if you just do not use the MPI error code (including `MPI_SUCCESS`) in `MPIDI_OFI_win_init_sep`.

2. Could you please also change the error returning mechanism in `MPIDI_OFI_win_init_stx` ? They need to be consistent.

shawnccx replied:
I'm thinking about using `MPIDI_OFI_SUCCESS` as return value for all those endpoint initialization function, to replace `MPI_SUCCESS`.

hajimefu replied:
@minsii I don't want to disturb the discussion too much, but can't we assume negative values won't conflict with MPI/MPICH error codes? Actually we are already doing similar thing here: https://github.com/pmodels/mpich/blob/master/src/mpid/ch4/netmod/ofi/ofi_send.h#L79

minsii replied:
@hajimefu : You are right that all MPI error codes are positive numbers as required in standard. So the negative error codes will not be conflict with any MPI error code.
Section 8.4 says: ```0 = MPI_SUCCESS < MPI_ERR...```

Perhaps we should at least add a comment on top of the function similar to `ofi_send.h`.

shawnccx replied:
I can add similar comments.

shawnccx commented:
Addressed comments, including: 
- add comment for return value of `MPIDI_win_init_` functions;
- apply error handling case to win_init_stx and per_win_sync; 
- close enpoint in fail case to release resource.

minsii commented on src/mpid/ch4/netmod/ofi/ofi_win.h:
```diff
@@ -111,6 +111,316 @@ static inline int MPIDI_OFI_win_allgather(MPIR_Win * win, void *base, int disp_u
+                        "falling back to global EP/counter scheme");
+            MPIDI_Global.rma_stx_ctx = NULL;
+            mpi_errno = MPI_ERR_OTHER;
+            goto fn_fail;
```
I guess here we want to return `MPIDI_OFI_ENAVAIL` ?

shawnccx commented:
The wrong return value for "failed STX init" has been fixed.

minsii commented:
Thanks for all the fixes ! I think the PR is ready to merge now. I will close it after last test.

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/ofi

shawnccx commented:
I thought `./spawn/spaiccreate2 2` is fixed by Gengbin's earlier PR.
